{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: transformers in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (4.39.3)\n",
      "Requirement already satisfied: filelock in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from transformers) (3.3.0)\n",
      "Requirement already satisfied: huggingface-hub<1.0,>=0.19.3 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from transformers) (0.22.2)\n",
      "Requirement already satisfied: numpy>=1.17 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from transformers) (1.26.2)\n",
      "Requirement already satisfied: packaging>=20.0 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from transformers) (21.3)\n",
      "Requirement already satisfied: pyyaml>=5.1 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from transformers) (6.0)\n",
      "Requirement already satisfied: regex!=2019.12.17 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from transformers) (2022.3.2)\n",
      "Requirement already satisfied: requests in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from transformers) (2.31.0)\n",
      "Requirement already satisfied: tokenizers<0.19,>=0.14 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from transformers) (0.15.2)\n",
      "Requirement already satisfied: safetensors>=0.4.1 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from transformers) (0.4.3)\n",
      "Requirement already satisfied: tqdm>=4.27 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from transformers) (4.62.3)\n",
      "Requirement already satisfied: fsspec>=2023.5.0 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from huggingface-hub<1.0,>=0.19.3->transformers) (2024.3.1)\n",
      "Requirement already satisfied: typing-extensions>=3.7.4.3 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from huggingface-hub<1.0,>=0.19.3->transformers) (4.10.0)\n",
      "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from packaging>=20.0->transformers) (2.4.7)\n",
      "Requirement already satisfied: colorama in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from tqdm>=4.27->transformers) (0.4.4)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from requests->transformers) (2.0.6)\n",
      "Requirement already satisfied: idna<4,>=2.5 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from requests->transformers) (3.2)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from requests->transformers) (1.25.11)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from requests->transformers) (2021.5.30)\n"
     ]
    }
   ],
   "source": [
    "! pip install transformers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pandas in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (1.3.3)\n",
      "Requirement already satisfied: nltk in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (3.7)\n",
      "Requirement already satisfied: mysql-connector-python in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (8.0.29)\n",
      "Requirement already satisfied: numpy>=1.17.3 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from pandas) (1.26.2)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from pandas) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2017.3 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from pandas) (2021.3)\n",
      "Requirement already satisfied: click in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from nltk) (7.1.2)\n",
      "Requirement already satisfied: joblib in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from nltk) (1.1.0)\n",
      "Requirement already satisfied: regex>=2021.8.3 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from nltk) (2022.3.2)\n",
      "Requirement already satisfied: tqdm in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from nltk) (4.62.3)\n",
      "Requirement already satisfied: protobuf>=3.0.0 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from mysql-connector-python) (4.23.4)\n",
      "Requirement already satisfied: six>=1.5 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from python-dateutil>=2.7.3->pandas) (1.15.0)\n",
      "Requirement already satisfied: colorama in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from tqdm->nltk) (0.4.4)\n"
     ]
    }
   ],
   "source": [
    "! pip install pandas nltk mysql-connector-python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mysql.connector\n",
    "import pandas as pd\n",
    "\n",
    "# Database connection parameters\n",
    "db_config = {\n",
    "    'host': 'localhost',\n",
    "    'user': 'root',\n",
    "    'password': 'Madgame@1',\n",
    "    'database': 'Bionexus_db'\n",
    "}\n",
    "\n",
    "# Connect to the database\n",
    "conn = mysql.connector.connect(**db_config)\n",
    "query = \"SELECT * FROM articles;\"  # Adjust this SQL query for your data\n",
    "df = pd.read_sql(query, conn)\n",
    "conn.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>authors</th>\n",
       "      <th>pub_date</th>\n",
       "      <th>journal</th>\n",
       "      <th>keywords</th>\n",
       "      <th>doi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>What is the role of microbial biotechnology an...</td>\n",
       "      <td>Microbial products are essential for developin...</td>\n",
       "      <td>Santos-Beneit Fernando</td>\n",
       "      <td>2024 Apr</td>\n",
       "      <td>MicrobiologyOpen</td>\n",
       "      <td>antibiotics, biotechnology, cancer, genetic en...</td>\n",
       "      <td>10.1002/mbo3.1406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>The prevalence of copy number increase at mult...</td>\n",
       "      <td>Copy number variation is a common contributor ...</td>\n",
       "      <td>Pokrovac Ivan, Rohner Nicolas, Pezer Željka</td>\n",
       "      <td>2024 Mar 31</td>\n",
       "      <td>Molecular ecology</td>\n",
       "      <td>cave colonization, copy number variation, gene...</td>\n",
       "      <td>10.1111/mec.17339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>Deceptive learning in histopathology.</td>\n",
       "      <td>Deep learning holds immense potential for hist...</td>\n",
       "      <td>Shahamatdar Sahar, Saeed-Vafa Daryoush, Linsle...</td>\n",
       "      <td>2024 Mar 31</td>\n",
       "      <td>Histopathology</td>\n",
       "      <td>KRAS, computational pathology, deep learning, ...</td>\n",
       "      <td>10.1111/his.15180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>Significant genomic introgression from grey ju...</td>\n",
       "      <td>Chicken is one of the most numerous and widely...</td>\n",
       "      <td>Zhao Xiurong, Wen Junhui, Zhang Xinye, Zhang J...</td>\n",
       "      <td>2024 Apr 01</td>\n",
       "      <td>Journal of animal science and biotechnology</td>\n",
       "      <td>BCO2, Domestic chickens, Grey junglefowl, Intr...</td>\n",
       "      <td>10.1186/s40104-024-01006-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>Transcriptomic changes in the microsporidia pr...</td>\n",
       "      <td>Congenital infection caused by vertical transm...</td>\n",
       "      <td>Shen Zigang, Ke Zhuojun, Yang Qiong, Ghebremic...</td>\n",
       "      <td>2024 Apr 01</td>\n",
       "      <td>BMC genomics</td>\n",
       "      <td>B. mori, N. Bombycis, Congenital infection, Ep...</td>\n",
       "      <td>10.1186/s12864-024-10236-y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59781</th>\n",
       "      <td>59782</td>\n",
       "      <td>Complex analysis of the national Hereditary an...</td>\n",
       "      <td>Hereditary angioedema (HAE) is a rare autosoma...</td>\n",
       "      <td>Markocsy Adam, Hrubiskova Katarina, Hrubisko M...</td>\n",
       "      <td>2024 Mar</td>\n",
       "      <td>The World Allergy Organization journal</td>\n",
       "      <td>Angioedemas, Complement C1 inhibitor protein, ...</td>\n",
       "      <td>10.1016/j.waojou.2024.100885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59782</th>\n",
       "      <td>59783</td>\n",
       "      <td>The Utility of a Novel Bidirectional Lateral I...</td>\n",
       "      <td>Smaller breasts require volume replacement to ...</td>\n",
       "      <td>Schwartz Jean-Claude D</td>\n",
       "      <td>2024 Mar</td>\n",
       "      <td>Plastic and reconstructive surgery. Global open</td>\n",
       "      <td></td>\n",
       "      <td>10.1097/GOX.0000000000005687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59783</th>\n",
       "      <td>59784</td>\n",
       "      <td>Navigating P2Y12 inhibition in the labyrinth o...</td>\n",
       "      <td>Cangrelor, a potent intravenous P2Y12 platelet...</td>\n",
       "      <td>Ali Abdelrahman, Jewani Poonam, Bourdillon Max...</td>\n",
       "      <td>2024</td>\n",
       "      <td>Frontiers in cardiovascular medicine</td>\n",
       "      <td>P2Y12 Inhibitors, bridging, cangrelor, coronar...</td>\n",
       "      <td>10.3389/fcvm.2024.1337957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59784</th>\n",
       "      <td>59785</td>\n",
       "      <td>Stalactites in the Right Ventricle.</td>\n",
       "      <td>Heart metastatic tumors are more frequent than...</td>\n",
       "      <td>Genovesi Eugenio, Di Marino Mario, Di Marco Ma...</td>\n",
       "      <td>2023 Oct-Dec</td>\n",
       "      <td>Journal of cardiovascular echography</td>\n",
       "      <td>Heart cancer, lung adenocarcinoma, multimodali...</td>\n",
       "      <td>10.4103/jcecho.jcecho_46_23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59785</th>\n",
       "      <td>59786</td>\n",
       "      <td>LncRNA HCP5-encoded protein contributes to adr...</td>\n",
       "      <td></td>\n",
       "      <td>Xing Jia-Ni, Shang Yi-Ni, Yu Zheng-Ling, Zhou ...</td>\n",
       "      <td>2024 Jul</td>\n",
       "      <td>Genes &amp; diseases</td>\n",
       "      <td></td>\n",
       "      <td>10.1016/j.gendis.2023.06.002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>59786 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          id                                              title  \\\n",
       "0          1  What is the role of microbial biotechnology an...   \n",
       "1          2  The prevalence of copy number increase at mult...   \n",
       "2          3              Deceptive learning in histopathology.   \n",
       "3          4  Significant genomic introgression from grey ju...   \n",
       "4          5  Transcriptomic changes in the microsporidia pr...   \n",
       "...      ...                                                ...   \n",
       "59781  59782  Complex analysis of the national Hereditary an...   \n",
       "59782  59783  The Utility of a Novel Bidirectional Lateral I...   \n",
       "59783  59784  Navigating P2Y12 inhibition in the labyrinth o...   \n",
       "59784  59785                Stalactites in the Right Ventricle.   \n",
       "59785  59786  LncRNA HCP5-encoded protein contributes to adr...   \n",
       "\n",
       "                                                abstract  \\\n",
       "0      Microbial products are essential for developin...   \n",
       "1      Copy number variation is a common contributor ...   \n",
       "2      Deep learning holds immense potential for hist...   \n",
       "3      Chicken is one of the most numerous and widely...   \n",
       "4      Congenital infection caused by vertical transm...   \n",
       "...                                                  ...   \n",
       "59781  Hereditary angioedema (HAE) is a rare autosoma...   \n",
       "59782  Smaller breasts require volume replacement to ...   \n",
       "59783  Cangrelor, a potent intravenous P2Y12 platelet...   \n",
       "59784  Heart metastatic tumors are more frequent than...   \n",
       "59785                                                      \n",
       "\n",
       "                                                 authors      pub_date  \\\n",
       "0                                 Santos-Beneit Fernando      2024 Apr   \n",
       "1            Pokrovac Ivan, Rohner Nicolas, Pezer Željka   2024 Mar 31   \n",
       "2      Shahamatdar Sahar, Saeed-Vafa Daryoush, Linsle...   2024 Mar 31   \n",
       "3      Zhao Xiurong, Wen Junhui, Zhang Xinye, Zhang J...   2024 Apr 01   \n",
       "4      Shen Zigang, Ke Zhuojun, Yang Qiong, Ghebremic...   2024 Apr 01   \n",
       "...                                                  ...           ...   \n",
       "59781  Markocsy Adam, Hrubiskova Katarina, Hrubisko M...      2024 Mar   \n",
       "59782                             Schwartz Jean-Claude D      2024 Mar   \n",
       "59783  Ali Abdelrahman, Jewani Poonam, Bourdillon Max...          2024   \n",
       "59784  Genovesi Eugenio, Di Marino Mario, Di Marco Ma...  2023 Oct-Dec   \n",
       "59785  Xing Jia-Ni, Shang Yi-Ni, Yu Zheng-Ling, Zhou ...      2024 Jul   \n",
       "\n",
       "                                               journal  \\\n",
       "0                                     MicrobiologyOpen   \n",
       "1                                    Molecular ecology   \n",
       "2                                       Histopathology   \n",
       "3          Journal of animal science and biotechnology   \n",
       "4                                         BMC genomics   \n",
       "...                                                ...   \n",
       "59781           The World Allergy Organization journal   \n",
       "59782  Plastic and reconstructive surgery. Global open   \n",
       "59783             Frontiers in cardiovascular medicine   \n",
       "59784             Journal of cardiovascular echography   \n",
       "59785                                 Genes & diseases   \n",
       "\n",
       "                                                keywords  \\\n",
       "0      antibiotics, biotechnology, cancer, genetic en...   \n",
       "1      cave colonization, copy number variation, gene...   \n",
       "2      KRAS, computational pathology, deep learning, ...   \n",
       "3      BCO2, Domestic chickens, Grey junglefowl, Intr...   \n",
       "4      B. mori, N. Bombycis, Congenital infection, Ep...   \n",
       "...                                                  ...   \n",
       "59781  Angioedemas, Complement C1 inhibitor protein, ...   \n",
       "59782                                                      \n",
       "59783  P2Y12 Inhibitors, bridging, cangrelor, coronar...   \n",
       "59784  Heart cancer, lung adenocarcinoma, multimodali...   \n",
       "59785                                                      \n",
       "\n",
       "                                doi  \n",
       "0                 10.1002/mbo3.1406  \n",
       "1                 10.1111/mec.17339  \n",
       "2                 10.1111/his.15180  \n",
       "3        10.1186/s40104-024-01006-7  \n",
       "4        10.1186/s12864-024-10236-y  \n",
       "...                             ...  \n",
       "59781  10.1016/j.waojou.2024.100885  \n",
       "59782  10.1097/GOX.0000000000005687  \n",
       "59783     10.3389/fcvm.2024.1337957  \n",
       "59784   10.4103/jcecho.jcecho_46_23  \n",
       "59785  10.1016/j.gendis.2023.06.002  \n",
       "\n",
       "[59786 rows x 8 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df = df.head(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import nltk\n",
    "from nltk.corpus import stopwords\n",
    "from nltk.stem import WordNetLemmatizer\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[nltk_data] Downloading package stopwords to\n",
      "[nltk_data]     C:\\Users\\spekande\\AppData\\Roaming\\nltk_data...\n",
      "[nltk_data]   Package stopwords is already up-to-date!\n",
      "[nltk_data] Downloading package wordnet to\n",
      "[nltk_data]     C:\\Users\\spekande\\AppData\\Roaming\\nltk_data...\n",
      "[nltk_data]   Package wordnet is already up-to-date!\n",
      "[nltk_data] Downloading package omw-1.4 to\n",
      "[nltk_data]     C:\\Users\\spekande\\AppData\\Roaming\\nltk_data...\n",
      "[nltk_data]   Package omw-1.4 is already up-to-date!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nltk.download('stopwords')\n",
    "nltk.download('wordnet')\n",
    "nltk.download('omw-1.4')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocess_text(text):\n",
    "    # Convert text to lowercase\n",
    "    text = text.lower()\n",
    "    # Remove non-alphabetic characters\n",
    "    text = re.sub(r'[^a-z\\s]', '', text)\n",
    "    # Tokenize text\n",
    "    tokens = nltk.word_tokenize(text)\n",
    "    # Remove stopwords\n",
    "    stop_words = set(stopwords.words('english'))\n",
    "    tokens = [token for token in tokens if token not in stop_words]\n",
    "    # Lemmatize\n",
    "    lemmatizer = WordNetLemmatizer()\n",
    "    tokens = [lemmatizer.lemmatize(token) for token in tokens]\n",
    "    # Re-join tokens into a string\n",
    "    return ' '.join(tokens)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\spekande\\AppData\\Local\\Temp/ipykernel_2424/4000351698.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  new_df['processed_abstract'] = new_df['abstract'].apply(preprocess_text)\n"
     ]
    }
   ],
   "source": [
    "new_df['processed_abstract'] = new_df['abstract'].apply(preprocess_text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>authors</th>\n",
       "      <th>pub_date</th>\n",
       "      <th>journal</th>\n",
       "      <th>keywords</th>\n",
       "      <th>doi</th>\n",
       "      <th>processed_abstract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>What is the role of microbial biotechnology an...</td>\n",
       "      <td>Microbial products are essential for developin...</td>\n",
       "      <td>Santos-Beneit Fernando</td>\n",
       "      <td>2024 Apr</td>\n",
       "      <td>MicrobiologyOpen</td>\n",
       "      <td>antibiotics, biotechnology, cancer, genetic en...</td>\n",
       "      <td>10.1002/mbo3.1406</td>\n",
       "      <td>microbial product essential developing various...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>The prevalence of copy number increase at mult...</td>\n",
       "      <td>Copy number variation is a common contributor ...</td>\n",
       "      <td>Pokrovac Ivan, Rohner Nicolas, Pezer Željka</td>\n",
       "      <td>2024 Mar 31</td>\n",
       "      <td>Molecular ecology</td>\n",
       "      <td>cave colonization, copy number variation, gene...</td>\n",
       "      <td>10.1111/mec.17339</td>\n",
       "      <td>copy number variation common contributor pheno...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>Deceptive learning in histopathology.</td>\n",
       "      <td>Deep learning holds immense potential for hist...</td>\n",
       "      <td>Shahamatdar Sahar, Saeed-Vafa Daryoush, Linsle...</td>\n",
       "      <td>2024 Mar 31</td>\n",
       "      <td>Histopathology</td>\n",
       "      <td>KRAS, computational pathology, deep learning, ...</td>\n",
       "      <td>10.1111/his.15180</td>\n",
       "      <td>deep learning hold immense potential histopath...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>Significant genomic introgression from grey ju...</td>\n",
       "      <td>Chicken is one of the most numerous and widely...</td>\n",
       "      <td>Zhao Xiurong, Wen Junhui, Zhang Xinye, Zhang J...</td>\n",
       "      <td>2024 Apr 01</td>\n",
       "      <td>Journal of animal science and biotechnology</td>\n",
       "      <td>BCO2, Domestic chickens, Grey junglefowl, Intr...</td>\n",
       "      <td>10.1186/s40104-024-01006-7</td>\n",
       "      <td>chicken one numerous widely distributed specie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>Transcriptomic changes in the microsporidia pr...</td>\n",
       "      <td>Congenital infection caused by vertical transm...</td>\n",
       "      <td>Shen Zigang, Ke Zhuojun, Yang Qiong, Ghebremic...</td>\n",
       "      <td>2024 Apr 01</td>\n",
       "      <td>BMC genomics</td>\n",
       "      <td>B. mori, N. Bombycis, Congenital infection, Ep...</td>\n",
       "      <td>10.1186/s12864-024-10236-y</td>\n",
       "      <td>congenital infection caused vertical transmiss...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>96</td>\n",
       "      <td>Targeting TGF-beta activation in cutaneous T-c...</td>\n",
       "      <td></td>\n",
       "      <td>Giustiniani Jérôme, Anh Ta Van, Belkhelouat Sa...</td>\n",
       "      <td>2024 Mar 28</td>\n",
       "      <td>The Journal of investigative dermatology</td>\n",
       "      <td>GARP, LRRC33, alphaV-beta1 integrin, microenvi...</td>\n",
       "      <td>10.1016/j.jid.2024.03.019</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>97</td>\n",
       "      <td>Single primer site-specific nested PCR for acc...</td>\n",
       "      <td>Genome-walking is a molecular tool used to unv...</td>\n",
       "      <td>Guo Xinyue, Zhu Yisong, Pan Zhenkang, Pan Hao,...</td>\n",
       "      <td>2024 Mar 28</td>\n",
       "      <td>Journal of microbiological methods</td>\n",
       "      <td>Genome-walking, Partial annealing, Single-prim...</td>\n",
       "      <td>10.1016/j.mimet.2024.106926</td>\n",
       "      <td>genomewalking molecular tool used unveil uncha...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>98</td>\n",
       "      <td>Oncogenic extrachromosomal DNA identification ...</td>\n",
       "      <td></td>\n",
       "      <td>Noorani Imran, Luebeck Jens, Rowan Andrew, Gro...</td>\n",
       "      <td>2024 Mar 28</td>\n",
       "      <td>Annals of oncology : official journal of the E...</td>\n",
       "      <td></td>\n",
       "      <td>10.1016/j.annonc.2024.03.008</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>99</td>\n",
       "      <td>Truncated variants of thyroid hormone receptor...</td>\n",
       "      <td>Thyroid hormone receptor β (THRβ) is a member ...</td>\n",
       "      <td>Rehman Ghausiya, Kashyap Jyoti, Srivastav Amit...</td>\n",
       "      <td>2024 Mar 28</td>\n",
       "      <td>Experimental cell research</td>\n",
       "      <td>Mitotic genome bookmarking, Resistance to thyr...</td>\n",
       "      <td>10.1016/j.yexcr.2024.114017</td>\n",
       "      <td>thyroid hormone receptor thr member nuclear re...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>100</td>\n",
       "      <td>Host genetic variants associated with COVID-19...</td>\n",
       "      <td>We present the results of an association study...</td>\n",
       "      <td>Skerenova Maria, Cibulka Michal, Dankova Zuzan...</td>\n",
       "      <td>2024 Mar 28</td>\n",
       "      <td>Advances in medical sciences</td>\n",
       "      <td>Haplotypes, Host genetic risk factors, Immunor...</td>\n",
       "      <td>10.1016/j.advms.2024.03.007</td>\n",
       "      <td>present result association study involving hos...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     id                                              title  \\\n",
       "0     1  What is the role of microbial biotechnology an...   \n",
       "1     2  The prevalence of copy number increase at mult...   \n",
       "2     3              Deceptive learning in histopathology.   \n",
       "3     4  Significant genomic introgression from grey ju...   \n",
       "4     5  Transcriptomic changes in the microsporidia pr...   \n",
       "..  ...                                                ...   \n",
       "95   96  Targeting TGF-beta activation in cutaneous T-c...   \n",
       "96   97  Single primer site-specific nested PCR for acc...   \n",
       "97   98  Oncogenic extrachromosomal DNA identification ...   \n",
       "98   99  Truncated variants of thyroid hormone receptor...   \n",
       "99  100  Host genetic variants associated with COVID-19...   \n",
       "\n",
       "                                             abstract  \\\n",
       "0   Microbial products are essential for developin...   \n",
       "1   Copy number variation is a common contributor ...   \n",
       "2   Deep learning holds immense potential for hist...   \n",
       "3   Chicken is one of the most numerous and widely...   \n",
       "4   Congenital infection caused by vertical transm...   \n",
       "..                                                ...   \n",
       "95                                                      \n",
       "96  Genome-walking is a molecular tool used to unv...   \n",
       "97                                                      \n",
       "98  Thyroid hormone receptor β (THRβ) is a member ...   \n",
       "99  We present the results of an association study...   \n",
       "\n",
       "                                              authors     pub_date  \\\n",
       "0                              Santos-Beneit Fernando     2024 Apr   \n",
       "1         Pokrovac Ivan, Rohner Nicolas, Pezer Željka  2024 Mar 31   \n",
       "2   Shahamatdar Sahar, Saeed-Vafa Daryoush, Linsle...  2024 Mar 31   \n",
       "3   Zhao Xiurong, Wen Junhui, Zhang Xinye, Zhang J...  2024 Apr 01   \n",
       "4   Shen Zigang, Ke Zhuojun, Yang Qiong, Ghebremic...  2024 Apr 01   \n",
       "..                                                ...          ...   \n",
       "95  Giustiniani Jérôme, Anh Ta Van, Belkhelouat Sa...  2024 Mar 28   \n",
       "96  Guo Xinyue, Zhu Yisong, Pan Zhenkang, Pan Hao,...  2024 Mar 28   \n",
       "97  Noorani Imran, Luebeck Jens, Rowan Andrew, Gro...  2024 Mar 28   \n",
       "98  Rehman Ghausiya, Kashyap Jyoti, Srivastav Amit...  2024 Mar 28   \n",
       "99  Skerenova Maria, Cibulka Michal, Dankova Zuzan...  2024 Mar 28   \n",
       "\n",
       "                                              journal  \\\n",
       "0                                    MicrobiologyOpen   \n",
       "1                                   Molecular ecology   \n",
       "2                                      Histopathology   \n",
       "3         Journal of animal science and biotechnology   \n",
       "4                                        BMC genomics   \n",
       "..                                                ...   \n",
       "95           The Journal of investigative dermatology   \n",
       "96                 Journal of microbiological methods   \n",
       "97  Annals of oncology : official journal of the E...   \n",
       "98                         Experimental cell research   \n",
       "99                       Advances in medical sciences   \n",
       "\n",
       "                                             keywords  \\\n",
       "0   antibiotics, biotechnology, cancer, genetic en...   \n",
       "1   cave colonization, copy number variation, gene...   \n",
       "2   KRAS, computational pathology, deep learning, ...   \n",
       "3   BCO2, Domestic chickens, Grey junglefowl, Intr...   \n",
       "4   B. mori, N. Bombycis, Congenital infection, Ep...   \n",
       "..                                                ...   \n",
       "95  GARP, LRRC33, alphaV-beta1 integrin, microenvi...   \n",
       "96  Genome-walking, Partial annealing, Single-prim...   \n",
       "97                                                      \n",
       "98  Mitotic genome bookmarking, Resistance to thyr...   \n",
       "99  Haplotypes, Host genetic risk factors, Immunor...   \n",
       "\n",
       "                             doi  \\\n",
       "0              10.1002/mbo3.1406   \n",
       "1              10.1111/mec.17339   \n",
       "2              10.1111/his.15180   \n",
       "3     10.1186/s40104-024-01006-7   \n",
       "4     10.1186/s12864-024-10236-y   \n",
       "..                           ...   \n",
       "95     10.1016/j.jid.2024.03.019   \n",
       "96   10.1016/j.mimet.2024.106926   \n",
       "97  10.1016/j.annonc.2024.03.008   \n",
       "98   10.1016/j.yexcr.2024.114017   \n",
       "99   10.1016/j.advms.2024.03.007   \n",
       "\n",
       "                                   processed_abstract  \n",
       "0   microbial product essential developing various...  \n",
       "1   copy number variation common contributor pheno...  \n",
       "2   deep learning hold immense potential histopath...  \n",
       "3   chicken one numerous widely distributed specie...  \n",
       "4   congenital infection caused vertical transmiss...  \n",
       "..                                                ...  \n",
       "95                                                     \n",
       "96  genomewalking molecular tool used unveil uncha...  \n",
       "97                                                     \n",
       "98  thyroid hormone receptor thr member nuclear re...  \n",
       "99  present result association study involving hos...  \n",
       "\n",
       "[100 rows x 9 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     microbial product essential developing various...\n",
       "1     copy number variation common contributor pheno...\n",
       "2     deep learning hold immense potential histopath...\n",
       "3     chicken one numerous widely distributed specie...\n",
       "4     congenital infection caused vertical transmiss...\n",
       "                            ...                        \n",
       "95                                                     \n",
       "96    genomewalking molecular tool used unveil uncha...\n",
       "97                                                     \n",
       "98    thyroid hormone receptor thr member nuclear re...\n",
       "99    present result association study involving hos...\n",
       "Name: processed_abstract, Length: 100, dtype: object"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df['processed_abstract']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: transformers in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (4.39.3)\n",
      "Requirement already satisfied: torch in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (2.2.2)\n",
      "Requirement already satisfied: pandas in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (1.3.3)\n",
      "Requirement already satisfied: filelock in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from transformers) (3.3.0)\n",
      "Requirement already satisfied: huggingface-hub<1.0,>=0.19.3 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from transformers) (0.22.2)\n",
      "Requirement already satisfied: numpy>=1.17 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from transformers) (1.26.2)\n",
      "Requirement already satisfied: packaging>=20.0 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from transformers) (21.3)\n",
      "Requirement already satisfied: pyyaml>=5.1 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from transformers) (6.0)\n",
      "Requirement already satisfied: regex!=2019.12.17 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from transformers) (2022.3.2)\n",
      "Requirement already satisfied: requests in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from transformers) (2.31.0)\n",
      "Requirement already satisfied: tokenizers<0.19,>=0.14 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from transformers) (0.15.2)\n",
      "Requirement already satisfied: safetensors>=0.4.1 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from transformers) (0.4.3)\n",
      "Requirement already satisfied: tqdm>=4.27 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from transformers) (4.62.3)\n",
      "Requirement already satisfied: typing-extensions>=4.8.0 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from torch) (4.10.0)\n",
      "Requirement already satisfied: sympy in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from torch) (1.12)\n",
      "Requirement already satisfied: networkx in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from torch) (3.2.1)\n",
      "Requirement already satisfied: jinja2 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from torch) (3.1.1)\n",
      "Requirement already satisfied: fsspec in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from torch) (2024.3.1)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from pandas) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2017.3 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from pandas) (2021.3)\n",
      "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from packaging>=20.0->transformers) (2.4.7)\n",
      "Requirement already satisfied: six>=1.5 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from python-dateutil>=2.7.3->pandas) (1.15.0)\n",
      "Requirement already satisfied: colorama in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from tqdm>=4.27->transformers) (0.4.4)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from jinja2->torch) (2.1.1)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from requests->transformers) (2.0.6)\n",
      "Requirement already satisfied: idna<4,>=2.5 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from requests->transformers) (3.2)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from requests->transformers) (1.25.11)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from requests->transformers) (2021.5.30)\n",
      "Requirement already satisfied: mpmath>=0.19 in c:\\users\\spekande\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.9_qbz5n2kfra8p0\\localcache\\local-packages\\python39\\site-packages (from sympy->torch) (1.3.0)\n"
     ]
    }
   ],
   "source": [
    "! pip install transformers torch pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "from transformers import AutoTokenizer, AutoModel\n",
    "import torch\n",
    "\n",
    "# Load BioBERT tokenizer and model\n",
    "tokenizer = AutoTokenizer.from_pretrained(\"dmis-lab/biobert-v1.1\")\n",
    "model = AutoModel.from_pretrained(\"dmis-lab/biobert-v1.1\")\n",
    "\n",
    "# Ensure a GPU is available for PyTorch\n",
    "if torch.cuda.is_available():\n",
    "    model = model.cuda()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "\n",
    "def generate_embeddings(text):\n",
    "    inputs = tokenizer(text, return_tensors=\"pt\", padding=True, truncation=True, max_length=512)\n",
    "    outputs = model(**inputs)\n",
    "    return outputs.last_hidden_state[:, 0, :].detach()\n",
    "    \n",
    "# Example: Generate embeddings for the first processed abstract\n",
    "sample_embedding = generate_embeddings(new_df['processed_abstract'][0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>authors</th>\n",
       "      <th>pub_date</th>\n",
       "      <th>journal</th>\n",
       "      <th>keywords</th>\n",
       "      <th>doi</th>\n",
       "      <th>processed_abstract</th>\n",
       "      <th>biobert_abstract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>What is the role of microbial biotechnology an...</td>\n",
       "      <td>Microbial products are essential for developin...</td>\n",
       "      <td>Santos-Beneit Fernando</td>\n",
       "      <td>2024 Apr</td>\n",
       "      <td>MicrobiologyOpen</td>\n",
       "      <td>antibiotics, biotechnology, cancer, genetic en...</td>\n",
       "      <td>10.1002/mbo3.1406</td>\n",
       "      <td>microbial product essential developing various...</td>\n",
       "      <td>[[0.028228626, 0.16243483, -0.37709862, -0.339...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>The prevalence of copy number increase at mult...</td>\n",
       "      <td>Copy number variation is a common contributor ...</td>\n",
       "      <td>Pokrovac Ivan, Rohner Nicolas, Pezer Željka</td>\n",
       "      <td>2024 Mar 31</td>\n",
       "      <td>Molecular ecology</td>\n",
       "      <td>cave colonization, copy number variation, gene...</td>\n",
       "      <td>10.1111/mec.17339</td>\n",
       "      <td>copy number variation common contributor pheno...</td>\n",
       "      <td>[[0.30040145, 0.11518507, -0.70688003, -0.3727...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>Deceptive learning in histopathology.</td>\n",
       "      <td>Deep learning holds immense potential for hist...</td>\n",
       "      <td>Shahamatdar Sahar, Saeed-Vafa Daryoush, Linsle...</td>\n",
       "      <td>2024 Mar 31</td>\n",
       "      <td>Histopathology</td>\n",
       "      <td>KRAS, computational pathology, deep learning, ...</td>\n",
       "      <td>10.1111/his.15180</td>\n",
       "      <td>deep learning hold immense potential histopath...</td>\n",
       "      <td>[[0.03513187, 0.019101435, -0.14304699, -0.046...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>Significant genomic introgression from grey ju...</td>\n",
       "      <td>Chicken is one of the most numerous and widely...</td>\n",
       "      <td>Zhao Xiurong, Wen Junhui, Zhang Xinye, Zhang J...</td>\n",
       "      <td>2024 Apr 01</td>\n",
       "      <td>Journal of animal science and biotechnology</td>\n",
       "      <td>BCO2, Domestic chickens, Grey junglefowl, Intr...</td>\n",
       "      <td>10.1186/s40104-024-01006-7</td>\n",
       "      <td>chicken one numerous widely distributed specie...</td>\n",
       "      <td>[[-0.00022922736, 0.121603325, -0.43926466, -0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>Transcriptomic changes in the microsporidia pr...</td>\n",
       "      <td>Congenital infection caused by vertical transm...</td>\n",
       "      <td>Shen Zigang, Ke Zhuojun, Yang Qiong, Ghebremic...</td>\n",
       "      <td>2024 Apr 01</td>\n",
       "      <td>BMC genomics</td>\n",
       "      <td>B. mori, N. Bombycis, Congenital infection, Ep...</td>\n",
       "      <td>10.1186/s12864-024-10236-y</td>\n",
       "      <td>congenital infection caused vertical transmiss...</td>\n",
       "      <td>[[0.1466351, 0.00940386, -0.6275529, -0.136923...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>96</td>\n",
       "      <td>Targeting TGF-beta activation in cutaneous T-c...</td>\n",
       "      <td></td>\n",
       "      <td>Giustiniani Jérôme, Anh Ta Van, Belkhelouat Sa...</td>\n",
       "      <td>2024 Mar 28</td>\n",
       "      <td>The Journal of investigative dermatology</td>\n",
       "      <td>GARP, LRRC33, alphaV-beta1 integrin, microenvi...</td>\n",
       "      <td>10.1016/j.jid.2024.03.019</td>\n",
       "      <td></td>\n",
       "      <td>[[0.13455093, -0.023871887, -0.17833501, -0.04...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>97</td>\n",
       "      <td>Single primer site-specific nested PCR for acc...</td>\n",
       "      <td>Genome-walking is a molecular tool used to unv...</td>\n",
       "      <td>Guo Xinyue, Zhu Yisong, Pan Zhenkang, Pan Hao,...</td>\n",
       "      <td>2024 Mar 28</td>\n",
       "      <td>Journal of microbiological methods</td>\n",
       "      <td>Genome-walking, Partial annealing, Single-prim...</td>\n",
       "      <td>10.1016/j.mimet.2024.106926</td>\n",
       "      <td>genomewalking molecular tool used unveil uncha...</td>\n",
       "      <td>[[0.11042314, 0.33067042, -0.3011033, -0.21766...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>98</td>\n",
       "      <td>Oncogenic extrachromosomal DNA identification ...</td>\n",
       "      <td></td>\n",
       "      <td>Noorani Imran, Luebeck Jens, Rowan Andrew, Gro...</td>\n",
       "      <td>2024 Mar 28</td>\n",
       "      <td>Annals of oncology : official journal of the E...</td>\n",
       "      <td></td>\n",
       "      <td>10.1016/j.annonc.2024.03.008</td>\n",
       "      <td></td>\n",
       "      <td>[[0.13455093, -0.023871887, -0.17833501, -0.04...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>99</td>\n",
       "      <td>Truncated variants of thyroid hormone receptor...</td>\n",
       "      <td>Thyroid hormone receptor β (THRβ) is a member ...</td>\n",
       "      <td>Rehman Ghausiya, Kashyap Jyoti, Srivastav Amit...</td>\n",
       "      <td>2024 Mar 28</td>\n",
       "      <td>Experimental cell research</td>\n",
       "      <td>Mitotic genome bookmarking, Resistance to thyr...</td>\n",
       "      <td>10.1016/j.yexcr.2024.114017</td>\n",
       "      <td>thyroid hormone receptor thr member nuclear re...</td>\n",
       "      <td>[[0.14231417, 0.023687487, -0.43822607, -0.121...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>100</td>\n",
       "      <td>Host genetic variants associated with COVID-19...</td>\n",
       "      <td>We present the results of an association study...</td>\n",
       "      <td>Skerenova Maria, Cibulka Michal, Dankova Zuzan...</td>\n",
       "      <td>2024 Mar 28</td>\n",
       "      <td>Advances in medical sciences</td>\n",
       "      <td>Haplotypes, Host genetic risk factors, Immunor...</td>\n",
       "      <td>10.1016/j.advms.2024.03.007</td>\n",
       "      <td>present result association study involving hos...</td>\n",
       "      <td>[[0.22211172, -0.032810032, -0.58305395, -0.20...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     id                                              title  \\\n",
       "0     1  What is the role of microbial biotechnology an...   \n",
       "1     2  The prevalence of copy number increase at mult...   \n",
       "2     3              Deceptive learning in histopathology.   \n",
       "3     4  Significant genomic introgression from grey ju...   \n",
       "4     5  Transcriptomic changes in the microsporidia pr...   \n",
       "..  ...                                                ...   \n",
       "95   96  Targeting TGF-beta activation in cutaneous T-c...   \n",
       "96   97  Single primer site-specific nested PCR for acc...   \n",
       "97   98  Oncogenic extrachromosomal DNA identification ...   \n",
       "98   99  Truncated variants of thyroid hormone receptor...   \n",
       "99  100  Host genetic variants associated with COVID-19...   \n",
       "\n",
       "                                             abstract  \\\n",
       "0   Microbial products are essential for developin...   \n",
       "1   Copy number variation is a common contributor ...   \n",
       "2   Deep learning holds immense potential for hist...   \n",
       "3   Chicken is one of the most numerous and widely...   \n",
       "4   Congenital infection caused by vertical transm...   \n",
       "..                                                ...   \n",
       "95                                                      \n",
       "96  Genome-walking is a molecular tool used to unv...   \n",
       "97                                                      \n",
       "98  Thyroid hormone receptor β (THRβ) is a member ...   \n",
       "99  We present the results of an association study...   \n",
       "\n",
       "                                              authors     pub_date  \\\n",
       "0                              Santos-Beneit Fernando     2024 Apr   \n",
       "1         Pokrovac Ivan, Rohner Nicolas, Pezer Željka  2024 Mar 31   \n",
       "2   Shahamatdar Sahar, Saeed-Vafa Daryoush, Linsle...  2024 Mar 31   \n",
       "3   Zhao Xiurong, Wen Junhui, Zhang Xinye, Zhang J...  2024 Apr 01   \n",
       "4   Shen Zigang, Ke Zhuojun, Yang Qiong, Ghebremic...  2024 Apr 01   \n",
       "..                                                ...          ...   \n",
       "95  Giustiniani Jérôme, Anh Ta Van, Belkhelouat Sa...  2024 Mar 28   \n",
       "96  Guo Xinyue, Zhu Yisong, Pan Zhenkang, Pan Hao,...  2024 Mar 28   \n",
       "97  Noorani Imran, Luebeck Jens, Rowan Andrew, Gro...  2024 Mar 28   \n",
       "98  Rehman Ghausiya, Kashyap Jyoti, Srivastav Amit...  2024 Mar 28   \n",
       "99  Skerenova Maria, Cibulka Michal, Dankova Zuzan...  2024 Mar 28   \n",
       "\n",
       "                                              journal  \\\n",
       "0                                    MicrobiologyOpen   \n",
       "1                                   Molecular ecology   \n",
       "2                                      Histopathology   \n",
       "3         Journal of animal science and biotechnology   \n",
       "4                                        BMC genomics   \n",
       "..                                                ...   \n",
       "95           The Journal of investigative dermatology   \n",
       "96                 Journal of microbiological methods   \n",
       "97  Annals of oncology : official journal of the E...   \n",
       "98                         Experimental cell research   \n",
       "99                       Advances in medical sciences   \n",
       "\n",
       "                                             keywords  \\\n",
       "0   antibiotics, biotechnology, cancer, genetic en...   \n",
       "1   cave colonization, copy number variation, gene...   \n",
       "2   KRAS, computational pathology, deep learning, ...   \n",
       "3   BCO2, Domestic chickens, Grey junglefowl, Intr...   \n",
       "4   B. mori, N. Bombycis, Congenital infection, Ep...   \n",
       "..                                                ...   \n",
       "95  GARP, LRRC33, alphaV-beta1 integrin, microenvi...   \n",
       "96  Genome-walking, Partial annealing, Single-prim...   \n",
       "97                                                      \n",
       "98  Mitotic genome bookmarking, Resistance to thyr...   \n",
       "99  Haplotypes, Host genetic risk factors, Immunor...   \n",
       "\n",
       "                             doi  \\\n",
       "0              10.1002/mbo3.1406   \n",
       "1              10.1111/mec.17339   \n",
       "2              10.1111/his.15180   \n",
       "3     10.1186/s40104-024-01006-7   \n",
       "4     10.1186/s12864-024-10236-y   \n",
       "..                           ...   \n",
       "95     10.1016/j.jid.2024.03.019   \n",
       "96   10.1016/j.mimet.2024.106926   \n",
       "97  10.1016/j.annonc.2024.03.008   \n",
       "98   10.1016/j.yexcr.2024.114017   \n",
       "99   10.1016/j.advms.2024.03.007   \n",
       "\n",
       "                                   processed_abstract  \\\n",
       "0   microbial product essential developing various...   \n",
       "1   copy number variation common contributor pheno...   \n",
       "2   deep learning hold immense potential histopath...   \n",
       "3   chicken one numerous widely distributed specie...   \n",
       "4   congenital infection caused vertical transmiss...   \n",
       "..                                                ...   \n",
       "95                                                      \n",
       "96  genomewalking molecular tool used unveil uncha...   \n",
       "97                                                      \n",
       "98  thyroid hormone receptor thr member nuclear re...   \n",
       "99  present result association study involving hos...   \n",
       "\n",
       "                                     biobert_abstract  \n",
       "0   [[0.028228626, 0.16243483, -0.37709862, -0.339...  \n",
       "1   [[0.30040145, 0.11518507, -0.70688003, -0.3727...  \n",
       "2   [[0.03513187, 0.019101435, -0.14304699, -0.046...  \n",
       "3   [[-0.00022922736, 0.121603325, -0.43926466, -0...  \n",
       "4   [[0.1466351, 0.00940386, -0.6275529, -0.136923...  \n",
       "..                                                ...  \n",
       "95  [[0.13455093, -0.023871887, -0.17833501, -0.04...  \n",
       "96  [[0.11042314, 0.33067042, -0.3011033, -0.21766...  \n",
       "97  [[0.13455093, -0.023871887, -0.17833501, -0.04...  \n",
       "98  [[0.14231417, 0.023687487, -0.43822607, -0.121...  \n",
       "99  [[0.22211172, -0.032810032, -0.58305395, -0.20...  \n",
       "\n",
       "[100 rows x 10 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "import torch.nn.functional as F\n",
    "\n",
    "def generate_embeddings(text):\n",
    "    inputs = tokenizer(text, return_tensors=\"pt\", padding=True, truncation=True, max_length=512)\n",
    "    outputs = model(**inputs)\n",
    "    # Extract the embeddings of the [CLS] token which is at position 0\n",
    "    return outputs.last_hidden_state[:, 0, :]\n",
    "\n",
    "# For the query, ensure it is a single tensor with the first dimension as 1\n",
    "query = \"cancer\"\n",
    "query_embedding = generate_embeddings(preprocess_text(query))\n",
    "\n",
    "# Stack all article embeddings into a tensor\n",
    "all_embeddings = torch.stack([generate_embeddings(text) for text in new_df['processed_abstract']])\n",
    "\n",
    "def find_similar_articles(query_embedding, all_embeddings, top_k=5):\n",
    "    # Compute cosine similarities\n",
    "    similarities = F.cosine_similarity(query_embedding, all_embeddings, dim=1)\n",
    "    # Get the indices of the top_k most similar articles\n",
    "    most_similar_indices = similarities.argsort(descending=True)[:top_k]\n",
    "    return most_similar_indices\n",
    "\n",
    "similar_articles_indices = find_similar_articles(query_embedding.unsqueeze(0), all_embeddings)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of articles in DataFrame: 100\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of articles in DataFrame:\", new_df.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: Each embedding must have the same number of elements as the query_embedding. Expected 1, got 768.\n"
     ]
    }
   ],
   "source": [
    "import torch\n",
    "\n",
    "def generate_embeddings(text):\n",
    "    # Placeholder function - replace this with your actual embedding generation logic.\n",
    "    # It should return a 1D tensor representing the embedding of the text.\n",
    "    # For example:\n",
    "    # return model.encode(text)  # Assuming 'model' is your loaded embedding model.\n",
    "    return torch.rand(768)  # Example: Random embedding vector of size 768.\n",
    "\n",
    "# Ensure 'new_df' contains your articles and 'processed_abstract' has the preprocessed text.\n",
    "embeddings_list = []\n",
    "for text in new_df['processed_abstract']:\n",
    "    embedding = generate_embeddings(text)\n",
    "    if embedding is None or not isinstance(embedding, torch.Tensor):\n",
    "        print(f\"Failed to generate embedding for text: {text}\")\n",
    "        continue  # Skip this entry or handle it as you see fit.\n",
    "    embeddings_list.append(embedding)\n",
    "\n",
    "if not embeddings_list:\n",
    "    print(\"No valid embeddings were generated.\")\n",
    "else:\n",
    "    all_embeddings = torch.stack(embeddings_list)  # Stack into a 2D tensor\n",
    "\n",
    "    # Example call to your function if 'query_embedding' is defined and all_embeddings is ready\n",
    "    # similar_articles_indices = find_similar_articles(query_embedding.unsqueeze(0), all_embeddings)\n",
    "\n",
    "# Validate the all_embeddings if created\n",
    "if 'all_embeddings' in locals():\n",
    "    if all_embeddings.dim() != 2:\n",
    "        print(\"Error: all_embeddings must be a 2D tensor where each row is an embedding.\")\n",
    "    elif all_embeddings.shape[1] != query_embedding.shape[0]:\n",
    "        print(f\"Error: Each embedding must have the same number of elements as the query_embedding. Expected {query_embedding.shape[0]}, got {all_embeddings.shape[1]}.\")\n",
    "\n",
    "# Further processing, such as finding similar articles, would go here.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_similar_articles(query_embedding, all_embeddings, top_k=5):\n",
    "    top_k = min(top_k, len(new_df))  # Adjust top_k to not exceed DataFrame size\n",
    "    similarities = F.cosine_similarity(query_embedding, all_embeddings, dim=1)\n",
    "    most_similar_indices = similarities.argsort(descending=True)[:top_k]\n",
    "    return most_similar_indices\n",
    "\n",
    "similar_articles_indices = find_similar_articles(query_embedding.unsqueeze(0), all_embeddings)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "positional indexers are out-of-bounds",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python39\\site-packages\\pandas\\core\\indexing.py\u001b[0m in \u001b[0;36m_get_list_axis\u001b[1;34m(self, key, axis)\u001b[0m\n\u001b[0;32m   1529\u001b[0m         \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1530\u001b[1;33m             \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_take_with_is_copy\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1531\u001b[0m         \u001b[1;32mexcept\u001b[0m \u001b[0mIndexError\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python39\\site-packages\\pandas\\core\\generic.py\u001b[0m in \u001b[0;36m_take_with_is_copy\u001b[1;34m(self, indices, axis)\u001b[0m\n\u001b[0;32m   3627\u001b[0m         \"\"\"\n\u001b[1;32m-> 3628\u001b[1;33m         \u001b[0mresult\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtake\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mindices\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mindices\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   3629\u001b[0m         \u001b[1;31m# Maybe set copy if we didn't actually change the index.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python39\\site-packages\\pandas\\core\\generic.py\u001b[0m in \u001b[0;36mtake\u001b[1;34m(self, indices, axis, is_copy, **kwargs)\u001b[0m\n\u001b[0;32m   3614\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 3615\u001b[1;33m         new_data = self._mgr.take(\n\u001b[0m\u001b[0;32m   3616\u001b[0m             \u001b[0mindices\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_get_block_manager_axis\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mverify\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python39\\site-packages\\pandas\\core\\internals\\managers.py\u001b[0m in \u001b[0;36mtake\u001b[1;34m(self, indexer, axis, verify)\u001b[0m\n\u001b[0;32m    861\u001b[0m         \u001b[0mn\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0maxis\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 862\u001b[1;33m         \u001b[0mindexer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mmaybe_convert_indices\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mn\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mverify\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mverify\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    863\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python39\\site-packages\\pandas\\core\\indexers.py\u001b[0m in \u001b[0;36mmaybe_convert_indices\u001b[1;34m(indices, n, verify)\u001b[0m\n\u001b[0;32m    291\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mmask\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0many\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 292\u001b[1;33m             \u001b[1;32mraise\u001b[0m \u001b[0mIndexError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"indices are out-of-bounds\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    293\u001b[0m     \u001b[1;32mreturn\u001b[0m \u001b[0mindices\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mIndexError\u001b[0m: indices are out-of-bounds",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[1;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_2424/2205666404.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[1;31m# Check if the list of indices is non-empty\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[1;32mif\u001b[0m \u001b[0msimilar_articles_indices\u001b[0m\u001b[1;33m:\u001b[0m  \u001b[1;31m# This checks if there are any indices in the list\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 6\u001b[1;33m     \u001b[0mrecommended_articles\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mnew_df\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0msimilar_articles_indices\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      7\u001b[0m     \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"Recommended Articles:\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      8\u001b[0m     \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mrecommended_articles\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'title'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'abstract'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python39\\site-packages\\pandas\\core\\indexing.py\u001b[0m in \u001b[0;36m__getitem__\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m    929\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    930\u001b[0m             \u001b[0mmaybe_callable\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mcom\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mapply_if_callable\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 931\u001b[1;33m             \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_getitem_axis\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmaybe_callable\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    932\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    933\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_is_scalar_access\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mtuple\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python39\\site-packages\\pandas\\core\\indexing.py\u001b[0m in \u001b[0;36m_getitem_axis\u001b[1;34m(self, key, axis)\u001b[0m\n\u001b[0;32m   1555\u001b[0m         \u001b[1;31m# a list of integers\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1556\u001b[0m         \u001b[1;32melif\u001b[0m \u001b[0mis_list_like_indexer\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1557\u001b[1;33m             \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_get_list_axis\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1558\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1559\u001b[0m         \u001b[1;31m# a single integer\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python39\\site-packages\\pandas\\core\\indexing.py\u001b[0m in \u001b[0;36m_get_list_axis\u001b[1;34m(self, key, axis)\u001b[0m\n\u001b[0;32m   1531\u001b[0m         \u001b[1;32mexcept\u001b[0m \u001b[0mIndexError\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1532\u001b[0m             \u001b[1;31m# re-raise with different error message\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1533\u001b[1;33m             \u001b[1;32mraise\u001b[0m \u001b[0mIndexError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"positional indexers are out-of-bounds\"\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[0merr\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1534\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1535\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_getitem_axis\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mint\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mIndexError\u001b[0m: positional indexers are out-of-bounds"
     ]
    }
   ],
   "source": [
    "# Convert tensor indices to a list for easier handling\n",
    "similar_articles_indices = find_similar_articles(query_embedding.unsqueeze(0), all_embeddings).tolist()\n",
    "\n",
    "# Check if the list of indices is non-empty\n",
    "if similar_articles_indices:  # This checks if there are any indices in the list\n",
    "    recommended_articles = new_df.iloc[similar_articles_indices]\n",
    "    print(\"Recommended Articles:\")\n",
    "    print(recommended_articles[['title', 'abstract']])\n",
    "else:\n",
    "    print(\"No articles found.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Recommended Articles:\n",
      "\n",
      "--- Article ---\n",
      "Title: Glucocorticoid receptor and specificity protein 1 (Sp1) or Sp3, but not the antibiotic Mithramycin A, stimulates human alphaherpesvirus 1 (HSV-1) replication.\n",
      "Abstract: Following acute human alphaherpesvirus 1 (HSV-1) infection of oral-facial mucosal surfaces, sensory neurons in trigeminal ganglia (TG) are important sites for life-long latency. Neurons in the central nervous system, including brainstem, also harbor viral genomes during latency. Periodically, certain cellular stressors trigger reactivation from latency, which can lead to recurrent HSV-1 disease: herpes labialis, herpes stromal keratitis, and encephalitis for example. Activation of the glucocorticoid receptor (GR) by stressful stimuli enhances HSV-1 gene expression, replication, and explant-induced reactivation. GR and certain stress-induced Krüppel like factors (KLF) cooperatively transactivate cis-regulatory modules (CRM) that drive expression of viral transcriptional regulatory proteins (ICP0, ICP4, and ICP27). These CRMs lack GR response elements (GRE); however, specificity protein 1 (Sp1) binding sites are crucial for GR and KLF15 or KLF4 mediated transactivation. Hence, we tested whether Sp1 or Sp3 regulate viral replication and transactivation of the ICP0 promoter. During early stages of explant-induced reactivation from latency, the number of Sp3+ TG neurons were significantly higher relative to TG from latently infected mice. Conversely, Sp1+ TG neurons were only increased in females, but not male mice, during explant-induced reactivation. Sp1 siRNA significantly reduced HSV-1 replication in cultured mouse (Neuro-2A) and monkey (CV-1) cells. Mithramycin A, an antibiotic that has anti-tumor activity preferentially interacts with GC-rich DNA, including Sp1 binding sites, significantly reduced HSV-1 replication indicating it has antiviral activity. GR and Sp1 or Sp3 transactivated the HSV-1 ICP0 promoter in Neuro-2A and CV-1 cells confirming these transcription factors enhance viral replication and gene expression.\n",
      "\n",
      "--- Article ---\n",
      "Title: Dynamics of viral DNA shedding and culture viral DNA positivity in different clinical samples collected during the 2022 mpox outbreak in Lombardy, Italy.\n",
      "Abstract: Mpox virus (MPXV) has recently spread outside of sub-Saharan Africa. This large multicentre study was conducted in Lombardy, the most densely populated Italian region accounting for more than 40% of Italian cases. The present study aims to: i) evaluate the presence and the shedding duration of MPXV DNA in different body compartments correlating the MPXV viability with the time to onset of symptoms; ii) provide evidence of MPXV persistence in different body compartment as a source of infection and iii) characterize the MPXV evolution by whole genome sequencing (WGS) during the outbreak occurred in Italy.\n",
      "\n",
      "--- Article ---\n",
      "Title: Recent advances in CRISPR-based functional genomics for the study of disease-associated genetic variants.\n",
      "Abstract: Advances in sequencing technology have greatly increased our ability to gather genomic data, yet understanding the impact of genetic mutations, particularly variants of uncertain significance (VUSs), remains a challenge in precision medicine. The CRISPR‒Cas system has emerged as a pivotal tool for genome engineering, enabling the precise incorporation of specific genetic variations, including VUSs, into DNA to facilitate their functional characterization. Additionally, the integration of CRISPR‒Cas technology with sequencing tools allows the high-throughput evaluation of mutations, transforming uncertain genetic data into actionable insights. This allows researchers to comprehensively study the functional consequences of point mutations, paving the way for enhanced understanding and increasing application to precision medicine. This review summarizes the current genome editing tools utilizing CRISPR‒Cas systems and their combination with sequencing tools for functional genomics, with a focus on point mutations.\n",
      "\n",
      "--- Article ---\n",
      "Title: [Clinicopathological and molecular genetic features of Crohn's disease].\n",
      "Abstract: No abstract available.\n",
      "\n",
      "--- Article ---\n",
      "Title: Engineering a Human P2X2 Receptor with Altered Ligand Selectivity in Yeast.\n",
      "Abstract: P2X receptors are a family of ligand gated ion channels found in a range of eukaryotic species including humans but are not naturally present in the yeast Saccharomyces cerevisiae. We demonstrate the first recombinant expression and functional gating of the P2X2 receptor in baker's yeast. We leverage the yeast host for facile genetic screens of mutant P2X2 by performing site saturation mutagenesis at residues of interest, including SNPs implicated in deafness and at residues involved in native binding. Deep mutational analysis and rounds of genetic engineering yield mutant P2X2 F303Y A304W, which has altered ligand selectivity towards the ATP analog AMP-PNP. The F303Y A304 variant shows over 100-fold increased intracellular calcium amplitudes with AMP-PNP compared to the WT receptor and has a much lower desensitization rate. Since AMP-PNP does not naturally activate P2X receptors, the F303Y A304 P2X2 may be a starting point for downstream applications in chemogenetic cellular control. Interestingly, the A304W mutation selectively destabilizes the desensitized state, which may provide a mechanistic basis for receptor opening with suboptimal agonists. The yeast system represents an inexpensive, scalable platform for ion channel characterization and engineering by circumventing the more expensive and time-consuming methodologies involving mammalian hosts.\n"
     ]
    }
   ],
   "source": [
    "def display_recommended_articles(articles):\n",
    "    if not articles.empty:\n",
    "        print(\"Recommended Articles:\")\n",
    "        for index, row in articles.iterrows():\n",
    "            print(\"\\n--- Article ---\")\n",
    "            print(\"Title:\", row['title'])\n",
    "            print(\"Abstract:\", row['abstract'] if row['abstract'] else \"No abstract available.\")\n",
    "    else:\n",
    "        print(\"No articles found.\")\n",
    "\n",
    "# Assuming similar_articles_indices is already defined and used to filter the DataFrame\n",
    "similar_articles_indices = find_similar_articles(query_embedding.unsqueeze(0), all_embeddings).tolist()\n",
    "recommended_articles = new_df.iloc[similar_articles_indices]\n",
    "\n",
    "# Display the formatted articles\n",
    "display_recommended_articles(recommended_articles)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Testing the above code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Recommended Articles:\n",
      "\n",
      "--- Article ---\n",
      "Title: Breast Cancer Treatment Innovations\n",
      "Abstract: New treatments for breast cancer are becoming more effective and less invasive.\n",
      "\n",
      "--- Article ---\n",
      "Title: Advances in Breast Cancer Research\n",
      "Abstract: This article discusses recent advances in the understanding of breast cancer.\n",
      "\n",
      "--- Article ---\n",
      "Title: Understanding Cellular Mechanisms in Lung Cancer\n",
      "Abstract: Lung cancer mechanisms are complex and require detailed study.\n",
      "\n",
      "--- Article ---\n",
      "Title: Molecular Pathways in Breast Cancer\n",
      "Abstract: Breast cancer involves specific molecular pathways that differ from other cancers.\n",
      "\n",
      "--- Article ---\n",
      "Title: Genetics of Skin Cancer\n",
      "Abstract: The genetic makeup of skin cancer reveals potential targets for therapy.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import torch\n",
    "\n",
    "# Mock DataFrame simulating your 'new_df' with titles and abstracts\n",
    "data = {\n",
    "    'title': [\n",
    "        \"Advances in Breast Cancer Research\",\n",
    "        \"Understanding Cellular Mechanisms in Lung Cancer\",\n",
    "        \"Breast Cancer Treatment Innovations\",\n",
    "        \"Genetics of Skin Cancer\",\n",
    "        \"Molecular Pathways in Breast Cancer\"\n",
    "    ],\n",
    "    'abstract': [\n",
    "        \"This article discusses recent advances in the understanding of breast cancer.\",\n",
    "        \"Lung cancer mechanisms are complex and require detailed study.\",\n",
    "        \"New treatments for breast cancer are becoming more effective and less invasive.\",\n",
    "        \"The genetic makeup of skin cancer reveals potential targets for therapy.\",\n",
    "        \"Breast cancer involves specific molecular pathways that differ from other cancers.\"\n",
    "    ]\n",
    "}\n",
    "\n",
    "new_df = pd.DataFrame(data)\n",
    "\n",
    "# Assume a preprocess function (simplified here) and generate mock embeddings\n",
    "def preprocess_text(text):\n",
    "    # Simplified preprocessing for demonstration\n",
    "    return ' '.join(word.lower() for word in text.split() if len(word) > 3)\n",
    "\n",
    "def mock_generate_embeddings(text):\n",
    "    # Generating random vectors to simulate embeddings\n",
    "    np.random.seed(sum(ord(char) for char in text))  # Seed for reproducibility based on text\n",
    "    return torch.tensor(np.random.rand(1, 768))  # Size of typical BERT embeddings\n",
    "\n",
    "# Simulate the entire process for the query \"breast cancer\"\n",
    "query = \"breast cancer\"\n",
    "processed_query = preprocess_text(query)\n",
    "query_embedding = mock_generate_embeddings(processed_query)\n",
    "\n",
    "# Simulate generating embeddings for all abstracts in the DataFrame\n",
    "all_embeddings = torch.cat([mock_generate_embeddings(preprocess_text(abstract)) for abstract in new_df['abstract']])\n",
    "\n",
    "# Function to find similar articles (mocked similarity calculation using dot product)\n",
    "def find_similar_articles(query_embedding, all_embeddings, top_k=5):\n",
    "    similarities = torch.mm(query_embedding, all_embeddings.t())  # Mock similarity\n",
    "    most_similar_indices = similarities.argsort(descending=True)[0][:top_k]\n",
    "    return most_similar_indices\n",
    "\n",
    "similar_articles_indices = find_similar_articles(query_embedding, all_embeddings)\n",
    "\n",
    "# Function to display recommended articles\n",
    "def display_recommended_articles(articles):\n",
    "    if not articles.empty:\n",
    "        print(\"Recommended Articles:\")\n",
    "        for index, row in articles.iterrows():\n",
    "            print(\"\\n--- Article ---\")\n",
    "            print(\"Title:\", row['title'])\n",
    "            print(\"Abstract:\", row['abstract'] if row['abstract'] else \"No abstract available.\")\n",
    "    else:\n",
    "        print(\"No articles found.\")\n",
    "\n",
    "# Retrieve and display the recommended articles\n",
    "recommended_articles = new_df.iloc[similar_articles_indices.tolist()]\n",
    "display_recommended_articles(recommended_articles)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
